pyadjoint/overloaded_type.py

```diff
@@ -240,6 +253,31 @@ def _ad_will_add_as_output(self):
         """
         return True
 
+    def _ad_to_petsc(self):
```

Comment:
@Ig-dolci I think this should have an optional `vec` argument, which will be set (and returned) if provided.

Reply:
I see. In fact, `_ad_to_petsc` with `vec` optional argument will be more useful.

Reply:
I just rewrote `_ad_to_petsc` with `vec` as an optional argument in  [PR 3657](https://github.com/firedrakeproject/firedrake/pull/3657).